atlasdb-client/src/main/java/com/palantir/atlasdb/keyvalue/impl/ProfilingKeyValueService.java

```diff
@@ -84,6 +86,34 @@ public static ProfilingKeyValueService create(KeyValueService delegate, long slo
         void log(String fmt, Object... args);
     }
 
+    interface FlushableLoggingFunction extends LoggingFunction {
+        void flush();
+    }
+
+    // Accumulates logs in a single string.
+    // Warning to users of this class: We do not guarantee that SLF4J special characters work properly across log lines,
+    // nor do we guarantee behaviour when the number of arguments does not match the number of placeholders.
+    @VisibleForTesting
+    static class LogAccumulator implements FlushableLoggingFunction {
+        private final StringBuilder combinedFormat = new StringBuilder();
+        private final List<Object> argList = Lists.newArrayList();
+        private final LoggingFunction sink;
+
+        LogAccumulator(LoggingFunction sink) {
+            this.sink = sink;
+        }
+
+        @Override
+        public synchronized void flush() {
+            sink.log(combinedFormat.toString(), argList.toArray(new Object[argList.size()]));
+        }
+
+        @Override
+        public synchronized void log(String fmt, Object... args) {
+            combinedFormat.append(fmt);
```

Comment:
do we need to include a new line or other delimiter between each accumulated message so that they're readable? new line should be ok as the accumulated group will still be logged as one logical blob for correlation

Reply:
I did think about that - the existing examples are okay, but that would make sense as well. I'll add `\n` between the appends then.